doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+- Element: `ElektraKey`
+- Name of element: `ElektraKeyname`
+- Value of element: `ElektraKeyvalue`
+
+Another (maybe less confusing) option would be:
+
+- Collection: `ElektraSet` or `ElektraMap`
+- Element: `ElektraEntry`
+- Name of element: `ElektraKey` or `ElektraName`
+- Value of element: `ElektraValue`
+
+### Note on public structs
+
+The public structs `ElektraKeyname` and `ElektraKeyvalue` defined below are not problematic or limiting in terms of forward compatibility.
+This is because, they essentially define what a "keyname" and a "key value" are for Elektra.
+That is why they are used both as part of the public API **and** in the definitions of the `*Cow` structs, which define how stuff is actually stored.
+Any change to these structs would mean changing that definition and therefore would be major breaking change, even if the structs where not public.
+
+### Public headers
+
+```c
+// indicates special rules for constness, see comments at use-site for details
+#define ELEKTRA_CONST const
+
+typedef uint8_t ElektraNamespace;
+
+typedef struct {
+    // constness follows containing struct
+    // if the containing struct is a `const ElektraKeyname *` then this field must not be modified
+    // if the containing struct is a `ElektraKeyname *` or a non-pointer `ElektraKeyname` it is okay to modify
+    ELEKTRA_CONST char * name;
+    size_t size;
+    ElektraNamespace ns;
+} ElektraKeyname;
+
+typedef struct {
+    const void * value;
+    size_t size;
+} ElektraKeyvalue;
+
+typedef struct ElektraKey ElektraKey;
+typedef struct ElektraKeyset ElektraKeyset;
+
+// 0 always indicates success, other values depend on function
+typedef int ElektraErrorCode;
+
+// Note: breaking change compared to current values
+typedef enum {
+    ELEKTRA_KEY_CP_NAME   = 1<<0,
+    ELEKTRA_KEY_CP_VALUE  = 1<<1,
+    ELEKTRA_KEY_CP_META   = 1<<2,
+    ELEKTRA_KEY_CP_ALL    = ELEKTRA_KEY_CP_NAME | ELEKTRA_KEY_CP_VALUE | ELEKTRA_KEY_CP_META,
+} ElektraKeyCopyFlag;
+
+void elektraKeynamePushPart (ElektraKeyname * name, const char * part);
+void elektraKeynamePopPart (ElektraKeyname * name);
+
+ElektraKey * elektraKeyNew (const ElektraKeyname * name);
+ElektraKey * elektraKeyRetain (const ElektraKey * key);
+void elektraKeyRelease (const ElektraKey * key);
+
+const ElektraKeyname * elektraKeyGetName (const ElektraKey * key);
+// Note: creates copy of name
+ElektraErrorCode elektraKeySetName (ElektraKey * key, const ElektraKeyname * name);
+// increments nameLock
+void elektraKeyLockName (ElektraKey * key);
+// decrements nameLock
+void elektraKeyUnlockName (ElektraKey * key);
+bool elektraKeyIsNameLocked (const ElektraKey * key);
+
+const ElektraKeyvalue * elektraKeyGetValue (const ElektraKey * key);
+// Note: creates copy of value
+ElektraErrorCode elektraKeySetValue (ElektraKey * key, const ElektraKeyvalue * value);
+
+/**
+ * @returns a `const ElektraKeyset *` that can safely be cast to `ElektraKeyset *`, iff a non-const `ElektraKey *` was passed as @p key
```

Comment:
?

Reply:
The constness works like for e.g. `strchr`:

```c
const char * cs;
char * s;

// OKAY
const char * x = strchr(cs);
char * x = strchr(s);
const char * x = strchr(s);

// ILLEGAL CAST, non-const pointer into const string
char * x = strchr(cs);
```

Basically what it means that, this is correct API use:

```c
ElektraKey * key;

ElektraKeyset * meta = (ElektraKeyset *) elektraKeyGetMeta (key);
```

But this is still wrong, because the metadata of a const `Key` should also be const:

```c
const ElektraKey * constKey;

ElektraKeyset * meta = (ElektraKeyset *) elektraKeyGetMeta (constKey);
```

> **Note** I didn't go with the "const arg, non-const return" variant from `strchr`, because the illegal case with `strchr` is not detected by the compiler (the illegal cast would actually be within `strchr`). With `elektraKeyGetMeta` the caller needs to make the cast explicitly otherwise there will be a compiler warning/error. Also I believe, the main reason why `strchr` uses `const` the way it does, is because the original API was written before `const` existed in C and changing it now would be a breaking change.